### PR TITLE
fix: make OneCycleLR pickleable

### LINKED ISSUE #94450: Pickling OneCycleLR.state_dict() with an unpickleable optimizer will result in an error.
### 🐛 Describe the bug

OneCycleLR.state_dict() returns a bound method of OneCycleLR. Pickling the state_dict() also pickles the optimizer object attached to the OneCycleLR class instance. This can result in a pickling fail if the attached optimizer itself isn't pickleable.

gist can be found here: https://gist.github.com/MikhailKardash/69c8e98c0e23dc01c99627a43a84981d

### Versions

PyTorch version: 1.9.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.8.15 (default, Nov 24 2022, 15:19:38)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.79.1-microsoft-standard-WSL2-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA T1200 Laptop GPU
Nvidia driver version: 517.13
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       GenuineIntel
Model name:                      11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
CPU family:                      6
Model:                           141
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        1
BogoMIPS:                        4992.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm avx512_vp2intersect flush_l1d arch_capabilities
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       384 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        10 MiB (8 instances)
L3 cache:                        24 MiB (1 instance)
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy==0.910
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==1.5.9
[pip3] torch==1.9.0
[pip3] torchaudio==0.13.1
[pip3] torchmetrics==0.11.0
[pip3] torchvision==0.10.0
[conda] _tflow_select             2.3.0                       mkl
[conda] blas                      1.0                         mkl
[conda] cpuonly                   2.0                           0    pytorch
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.23.5           py38h14f4228_0
[conda] numpy-base                1.23.5           py38h31eccc5_0
[conda] pytorch-lightning         1.5.9                    pypi_0    pypi
[conda] pytorch-mutex             1.0                         cpu    pytorch
[conda] torch                     1.9.0                    pypi_0    pypi
[conda] torchaudio                0.13.1                 py38_cpu    pytorch
[conda] torchmetrics              0.11.0                   pypi_0    pypi
[conda] torchvision               0.10.0                   pypi_0    pypi

cc @vincentqb @jbschlosser @albanD @janeyx99

### PR DESCRIPTION
**Pull request by MikhailKardash**
Fixes #94450

Drop `anneal_func` from the state_dict and reinitialize it when loading state_dict.

### CODE REVIEW

File: test/optim/test_lrscheduler.py [hunk 1]
```diff
@@ -1703,6 +1703,32 @@ def test_cycle_lr_with_adam(self):
...
+        pickle.dumps(state)
+
+    def test_onecycle_lr_restore_from_state_dict(self):
+        adam_opt = Adam(self.net.parameters())
+
+        # case 1: cosine annealing
+        scheduler = OneCycleLR(adam_opt, max_lr=5, total_steps=200, anneal_strategy='cos')
+        restored_scheduler = OneCycleLR(adam_opt, max_lr=5, total_steps=200)
+        restored_scheduler.load_state_dict(scheduler.state_dict())
+        self.assertTrue(restored_scheduler.anneal_strategy == scheduler.anneal_strategy == 'cos')
```
Comment (janeyx99) [inline test/optim/test_lrscheduler.py:1720]: Could we switch these to two assertEquals for better test error messaging if someone causes this test to fail in the future?

**Review by janeyx99 — COMMENTED** [review]
Hey, this looks awesome! Do you plan on finishing this PR? I've left one review item.

**Comment by MikhailKardash** [comment]
> Hey, this looks awesome! Do you plan on finishing this PR? I've left one review item.

Yeah, but I have to get a new CLA first. I'll open this PR once that's resolved.
In the meantime, I'll get your requested change in soon.

**Review by janeyx99 — APPROVED** [review]

**Comment by MikhailKardash** [comment]
https://github.com/pytorch/pytorch/issues/94450
given that the issue is closed, there's no longer a need for this PR.

**Pull request closed without merging**